### PR TITLE
Forward Authentication: add X-Forwarded-Uri

### PR DESCRIPTION
**Pull request by sebastianbauer**
### What does this PR do?

Added requested URI to auth server for Forward Authentication.

### Motivation

Missing feature for authentication decision based on the URI (frontend).

### More

Added implementation, docs and test.

### Additional Notes

Could be good enough for #2162 as well.

### CODE REVIEW

File: middlewares/auth/forward_test.go [hunk 1]
```diff
@@ -229,10 +257,9 @@ func Test_writeHeader(t *testing.T) {
 				req.Host = ""
 			}
 
-			forwardReq := testhelpers.MustNewRequest(http.MethodGet, "http://foo.bar", nil)
+			forwardReq := testhelpers.MustNewRequest(http.MethodGet, "http://foo.bar/path?q=1", nil)
 
 			writeHeader(req, forwardReq, test.trustForwardHeader)
-
```
Comment (ldez) [inline middlewares/auth/forward_test.go]: Could you restore this line?

Comment (sebastianbauer) [reply middlewares/auth/forward_test.go]: Of course. Sorry for this.

Can you tell me which tool will handle such formatting? `go fmt` does not. Or was it just manual checking? Thanks.

Comment (ldez) [reply middlewares/auth/forward_test.go]: It's `gofmt` on our CI: https://semaphoreci.com/containous/traefik/branches/pull-request-2398/builds/3

File: middlewares/auth/forward.go [hunk 2]
```diff
@@ -12,6 +12,10 @@ import (
 	"github.com/vulcand/oxy/utils"
 )
 
+const (
+	XForwardedURI = "X-Forwarded-Uri"
```
Comment (ldez) [inline middlewares/auth/forward.go]: Could you rename to `xForwardedURI`?

File: middlewares/auth/forward_test.go (hunk 1 above)
Comment (sebastianbauer) [reply middlewares/auth/forward_test.go]: Thank you. I see and will check all those validation tests you integrated. Sadly `validate-gofmt` doesn't producde the error when the blank line is missing (Go 1.9.2).

I will check those validation results on GitHub next time.

**Comment by mmatur** [comment]
@sebastianbauer Could you please rebase your PR on master  ?

**Pull request merged**